### ISSUE TITLE
Unable to add words to custom dictionary

### ISSUE THREAD

**Issue by memeplex** [streetsidesoftware/vscode-spell-checker#1930]
If I set

```json
  "cSpell.customDictionaries": {
    "local": {
      "name": "local",
      "path": "~/Library/Spelling/LocalDictionary",
      "addWords": true,
      "scope": "user",
    }
  },
```

for my user I only get the options to add words to user or workspace settings in the context menu or quick fix menu.

`~/Library/Spelling/LocalDictionary` is a file containing a list of words, one word per line.

Weirdly enough, with this config the option to add to the user dictionary appears in the menus:

```json
  "cSpell.customDictionaries": {
    "custom-dictionary-user": {
      "name": "custom-dictionary-user",
      "path": "~/.cspell/custom-dictionary-user.txt",
      "addWords": true,
      "scope": "user"
    }
}
```

Version 2.1.11

**Comment by memeplex** [comment]
Some things I found very confusing:
- The generated user dictionary `custom-dictionary-user` is not added to the `cSpell.dictionaries` list. Is this list necessary at all?
- Some examples in the documentation use `cSpell.dictionaryDefinitions` while others use `cSpell.customDictionaries`. The generated `custom-dictionary-user` also uses `cSpell.customDictionaries`. What is the difference?
- What is the `scope` setting for? I wasn't able to find any documentation about it.

**Comment by memeplex** [comment]
> I only get the options to add words to user or workspace settings

I think this is due to my dictionary not having the `.txt` extension. For example, with a `/tmp/local` dictionary no option to add to the user dictionary appears in the menu, but as soon as I change the file to `/tmp/local.txt` the option appears.

My use case is that I want to share my MacOS standard user dictionary in `~/Library/Spelling/LocalDictionary` with cSpell. I could just create a symlink with a `.txt` extension, but such limitation seems pretty arbitrary to me.

**Comment by DizzyHSlightlyVoided** [comment]
I had this exact problem when I tried to do the same thing with the Windows system spellcheck-dictionary, with another wrinkle added as well: Code Spell Checker does *not* like encodings other than UTF-8, and tried to interpret the UTF-16 .dic file (and insert text into it!) as UTF-8.

**Comment by Jason3S** [comment]
Note: Extensions have very little control over how VS Code Settings are presented.

A bit of background:
- `dictionaryDefinitions` - Is a list that allows you to define dictionaries that can be used.
- `dictionaries` - is a list that allows you to choose when they are used.
- `words` - list of words to be considered correct.
- `userWords` - list of words to be considered correct (defined at the user level).

VS Code settings do not play nice with lists. A list defined at the User level will be completely overwritten by a list of the same name at the Workspace or Folder level. Which means if a project defines a Workspace dictionary, you won't be able to use your User dictionary. This is also why `cSpell.words` and `cSpell.userWords` exist.

`cSpell.customDictionaries` was later added as a way to work around the VS Code list merge issue. VS Code will merge top level objects. 

## Example of how VS Code merges settings:

**Settings - User**
```json
"cSpell.customDictionaries": {
  "User Words": { "path": "~/my-words.txt" }
},
"cSpell.words": ["user"]
```

**Settings - Workspace**
```json
"cSpell.customDictionaries": {
  "Project Words": { "path": "${workspaceRoot}/project-words.txt" }
},
"cSpell.words": ["workspace"]
```

**Settings Used**
```json
"cSpell.customDictionaries": {
  "User Words": { "path": "~/my-words.txt" },
  "Project Words": { "path": "${workspaceRoot}/project-words.txt" }
},
"cSpell.words": ["workspace"]
```

## Q&A

> * The generated user dictionary `custom-dictionary-user` is not added to the `cSpell.dictionaries` list. Is this list necessary at all?

`customDictionaries` are automatically/silently added to the list of dictionaries. 

> * Some examples in the documentation use `cSpell.dictionaryDefinitions` while others use `cSpell.customDictionaries`. 
The generated `custom-dictionary-user` also uses `cSpell.customDictionaries`. What is the difference?

Sorry for the confusion, these exist to workaround limitations in VS Code. `cSpell.dictionaryDefinitions` allows you to define a list of available dictionaries, this matches CSpell command line tool's `dictionaryDefinitions`. Because it is not possible to merge lists in VS Code Settings, `cSpell.customDictionaries` was added as a workaround to that problem.

> * What is the `scope` setting for? I wasn't able to find any documentation about it.

`scope` can be `user`, `workspace`, or `folder`. These match the VS Code settings. This is used when asking the extension to add a word to your dictionary. If there is only one `user` dictionary, when you ask the spell checker to add a word to your User Words, then it will add it without needing to ask which dictionary.

**Comment by Jason3S** [comment]
@DizzyHSlightlyVoided 
> I had this exact problem when I tried to do the same thing with the Windows system spellcheck-dictionary, with another wrinkle added as well: Code Spell Checker does _not_ like encodings other than UTF-8, and tried to interpret the UTF-16 .dic file (and insert text into it!) as UTF-8.

The spell checker doesn't currently support UTF-16 dictionaries.
Would you be able to add your use case to [Support UTF-16 word lists · Issue #2805 · streetsidesoftware/cspell](https://github.com/streetsidesoftware/cspell/issues/2805)? It is not so clear why UTF-8 would be insufficient.

**Comment by DizzyHSlightlyVoided** [comment]
> @DizzyHSlightlyVoided
> 
> > I had this exact problem when I tried to do the same thing with the Windows system spellcheck-dictionary, with another wrinkle added as well: Code Spell Checker does _not_ like encodings other than UTF-8, and tried to interpret the UTF-16 .dic file (and insert text into it!) as UTF-8.
> 
> The spell checker doesn't currently support UTF-16 dictionaries. Would you be able to add your use case to [Support UTF-16 word lists · Issue #2805 · streetsidesoftware/cspell](https://github.com/streetsidesoftware/cspell/issues/2805)? It is not so clear why UTF-8 would be insufficient.

Got it, can do.

**Comment by Jason3S** [comment]
@memplex,

This was fixed by #2002
